Stratis.Bitcoin/BlockPulling/QualityScore.cs

```diff
@@ -0,0 +1,201 @@
+        /// <remarks>All access to this object has to be protected by <see cref="lockObject"/>.</remarks>
+        private readonly Dictionary<IBlockPullerBehavior, int> peerReferenceCounter;
+
+        /// <summary>Sum of all samples in <see cref="samples"/> array.</summary>
+        /// <remarks>All access to this object has to be protected by <see cref="lockObject"/>.</remarks>
+        private double samplesSum { get; set; }
+
+        /// <summary>
+        /// Initializes a new instance of the object.
+        /// </summary>
+        /// <param name="maxSampleCount">Maximal number of samples we calculate statistics from.</param>
+        /// <param name="loggerFactory">Factory to be used to create logger for the puller.</param>
+        public QualityScore(int maxSampleCount, ILoggerFactory loggerFactory)
+        {
+            this.logger = loggerFactory.CreateLogger(this.GetType().FullName);
+
+            this.samples = new PeerSample[maxSampleCount];
+            for (int i = 0; i < this.samples.Length; i++)
+                this.samples[i] = new PeerSample();
+
+            this.AverageBlockTimePerKb = 0.0;
+            this.peerReferenceCounter = new Dictionary<IBlockPullerBehavior, int>();
+        }
+
+        /// <summary>
+        /// Adds new time of a block to the list of times of recently downloaded blocks.
+        /// </summary>
+        /// <param name="peer">Peer that downloaded the block.</param>
+        /// <param name="blockDownloadTimeMs">Time in milliseconds it took to download the block from the peer.</param>
+        /// <param name="blockSize">Size of the downloaded block in bytes.</param>
+        public void AddSample(IBlockPullerBehavior peer, long blockDownloadTimeMs, int blockSize)
+        {
+            this.logger.LogTrace($"({nameof(peer)}:{peer.GetHashCode():x},{nameof(blockDownloadTimeMs)}:{blockDownloadTimeMs},{nameof(blockSize)}:{blockSize})");
+
+            double timePerKb = 1024.0 * (double)blockDownloadTimeMs / (double)blockSize;
+            if (timePerKb < 0.00001) timePerKb = 0.00001;
+
+            lock (this.lockObject)
+            {
+                // If we reached the maximum number of samples, we need to remove oldest sample.
+                if (this.samplesCount == this.samples.Length)
+                {
+                    PeerSample oldSample = this.samples[this.samplesIndex];
+                    this.samplesSum -= oldSample.timePerKb;
+                    this.peerReferenceCounter[oldSample.peer]--;
+
+                    if (this.peerReferenceCounter[oldSample.peer] == 0)
+                        this.peerReferenceCounter.Remove(oldSample.peer);
+                }
+                else this.samplesCount++;
+
+                // Add new sample to the mix.
+                this.samples[this.samplesIndex].timePerKb = timePerKb;
+                this.samples[this.samplesIndex].peer = peer;
+                this.samplesIndex = (this.samplesIndex + 1) % this.samples.Length;
```

Comment:
Ah this will reset back to zero so you start from the beginning of the samples array.
I wonder why not use a Queue and just pop the last entry, you will save a lot of logic?